### ISSUE TITLE
In Wayland, Solaar cannot find active process name causing Solaar to print a traceback and Rules do not work

### ISSUE THREAD

**Issue by lieryan** [pwr-Solaar/Solaar#1207]
Running Solaar in Wayland causes a lot of crashes that looks like so:

```
ERROR:logitech_receiver.listener:processing Notification(10,255,00,00,000000)
Traceback (most recent call last):
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/listener.py", line 193, in run
    self._notifications_callback(n)
  File "/home/lieryan/Projects/Solaar/lib/solaar/listener.py", line 190, in _notifications_handler
    _notifications.process(self.receiver, n)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/notifications.py", line 62, in process
    return _process_device_notification(device, status, notification)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/notifications.py", line 138, in _process_device_notification
    return _process_feature_notification(device, status, n, feature)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/notifications.py", line 388, in _process_feature_notification
    _diversion.process_notification(device, status, n, feature)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/diversion.py", line 792, in process_notification
    rules.evaluate(feature, notification, device, status, True)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/diversion.py", line 248, in evaluate
    result = component.evaluate(feature, notification, device, status, result)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/diversion.py", line 248, in evaluate
    result = component.evaluate(feature, notification, device, status, result)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/diversion.py", line 248, in evaluate
    result = component.evaluate(feature, notification, device, status, result)
  File "/home/lieryan/Projects/Solaar/lib/logitech_receiver/diversion.py", line 343, in evaluate
    return active_process_name.startswith(self.process) or active_wm_class_name.startswith(self.process)
AttributeError: 'NoneType' object has no attribute 'startswith'
```

Digging into diversion.py, it looks like Wayland never send `NET_ACTIVE_WINDOW`, so the global variables `active_process_name` and `active_wm_class_name` are never set.

Possibly because of the traceback, it looks like that the Rules system doesn't work at all.

**Comment by pfps** [comment]
Rules are known not to work in Wayland because of Waylands much stricter security stance.  This is documented.

However, Solaar should not crash when run under Wayland.

**Comment by lieryan** [comment]
@pfps Thanks for the quick turnaround on looking at this issue. From a quick testing, this seems to have silenced the traceback and doesn't seem to have any other adverse effect (other than Wayland just being usual Wayland and most Rules don't work):

```
M lib/logitech_receiver/diversion.py
@@ -340,7 +340,8 @@ class Process(Condition):
     def evaluate(self, feature, notification, device, status, last_result):
         if not isinstance(self.process, str):
             return False
-        return active_process_name.startswith(self.process) or active_wm_class_name.startswith(self.process)
+        return (active_process_name and active_process_name.startswith(self.process)) or \
+            (active_wm_class_name and active_wm_class_name.startswith(self.process))
 
     def data(self):
         return {'Process': str(self.process)}
```

It would probably be good to document (and indicate in the GUI) which rules does not work in Wayland. As I understand Rules that injects keyboard events doesn't seem to work on Wayland, because it lacks support for it; but I can see that there are some rules that do work, like Gestures Condition and Execute Action, which can still be used to make some useful actions.

**Comment by pfps** [comment]
Aah.  It looks as if you have some X11 libraries available while running Wayland and but your Process conditions are causing errors.  I hadn't thought of that situation.  I'm not sure what will work and what won't in this situation but there shouldn't be errors.

I've patched the code to eliminate the error you have encountered but there might be more.  I've also updated the documentation for rules.  So please try out PR #1208 and let me know whether you still get the same errors and what new errors you get.



To download and work with Solar
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1208, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1208/head:pull_1208
git checkout pull_1208
```
To download a new version of the pull request, fetch it and then set your pull branch
to the new fetch, as in:
```
git checkout pull_1208
git fetch origin pull/1208/head
git reset --hard FETCH_HEAD
```

**Comment by lieryan** [comment]
Hi @pfps, yes, I have both Wayland and X11 libraries installed. From my testing, the following works:

Condition:

1. Feature: Works great
2. Process: Don't work
3. Report: I don't know how this feature works
4. Modifiers: Don't work
5. Key: Don't work
6. Test: I don't know how this feature works
7. Mouse Gesture: Works great

Action:

1. Key press: Don't work
2. Mouse scroll: Don't work
3. Mouse click: Don't work
4. Execute: Works great

Seems like until Wayland implements the features necessary here for applications to grab and simulate inputs, the future of Rules on Wayland seems rather grim; that said, at least Mouse Gesture and Execute works, so all might not be lost. It may be able to bypass Wayland and grab input directly from uinput, but this will require Solaar to run as root.

**Comment by pfps** [comment]
OK.  Please run the PR and let me know whether you encounter any more errors.

**Comment by pfps** [comment]
Oh, I see that you did run it.

**Comment by pfps** [comment]
Fixed by PR #1208
